pymc3/distributions/distribution.py

```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
+
+        # Warn about discouraged parametrization
+        if shape is not None and not isinstance(shape, (list, tuple)):
+            warnings.warn("The `shape` parameter should be a list or tuple.", UserWarning)
+            shape = (shape,)
+        if dims is not None and not isinstance(dims, (list, tuple)):
+            warnings.warn("The `dims` parameter should be a list or tuple.", UserWarning)
+            dims = (dims,)
+        if size is not None and not isinstance(size, (list, tuple)):
+            warnings.warn("The `size` parameter should be a list or tuple.", UserWarning)
+            size = (size,)
+
+        if size is not None and Ellipsis in size:
+            raise ValueError("The `size` parameter cannot contain an Ellipsis. Actual: {size}")
+        if not _valid_ellipsis_position(shape):
+            raise ValueError(
+                f"Ellipsis in `shape` may only appear in the last position. Actual: {shape}"
+            )
+        if not _valid_ellipsis_position(dims):
+            raise ValueError(
+                f"Ellipsis in `dims` may only appear in the last position. Actual: {dims}"
+            )
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        # Create the RV without specifying size or testval.
+        # The size will be expanded later (if necessary) and only then the testval fits.
+        rv_native = cls.dist(*args, rng=rng, testval=None, size=None, **kwargs)
+        implied_ndim = rv_native.ndim
+        implied_dims = ...  # TODO: infer dimension names from Variables!
+
+        if shape is None and dims is None and size is None:
+            size = ()
+        elif size is not None:
+            # User already specified how to expand the RV
+            pass
+        elif shape is not None:
+            # Infer size from shape
+            if Ellipsis in shape:
+                size = tuple(shape[:-1])
+            else:
+                size = tuple(shape[: len(shape) - implied_ndim])
+                if size:
+                    warnings.warn(
+                        f"The specified shape {shape} has {len(shape)} dimensions. "
+                        f"This is more than the {implied_ndim} dimensions implied by the distribution parameters. "
+                        f"To replicate the RV beyond its implied dimensionality use `size={size}` instead.",
+                        UserWarning,
+                    )
+        elif dims is not None:
+            # Infer size from dims (and coords)
+            if Ellipsis in dims:
+                raise NotImplementedError("Ellipsis-based dims inference is not implemented.")
+                dims = tuple(dims[:-1]) + implied_dims
+            dimshape = tuple(len(model.coords[dname]) for dname in dims)
+            size = tuple(dimshape[: len(dimshape) - implied_ndim])
+        else:
+            raise Exception("This should have been unreachable code.")
```

Comment:
If this is truly unreachable, then it isn't necessary; otherwise, there's something missing in this logic that needs to be addressed directly